PowerEditor/src/WinControls/ProjectPanel/Directory.cpp

```diff
@@ -0,0 +1,454 @@
+		FindClose(hFind);
+
+	// no file was found in this directory. Now check the sub directories.
+	hFind = FindFirstFileEx(searchPath.c_str(), FindExInfoBasic, &fd, FindExSearchLimitToDirectories, NULL, 0);
+
+	if (hFind == INVALID_HANDLE_VALUE)
+		return false;
+
+	while (hFind != INVALID_HANDLE_VALUE)
+	{
+
+		// a directory was found
+		if ((fd.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY) != 0)
+		{
+			file = fd.cFileName;
+			// skip . and ..
+			if (file == TEXT(".") || file == TEXT(".."))
+			{
+				if (!FindNextFile(hFind, &fd))
+					break;
+				continue;
+			}
+
+			// check if this child directory contains data; if this is the case, return true
+			childPath = path + TEXT("\\") + file;
+			if (containsData(childPath))
+			{
+				FindClose(hFind);
+				return true;
+			}
+		}
+		else
+		{
+			// data was found - return true. This should never happen, because the files were already checked. Just to be safe.
+			FindClose(hFind);
+			return true;
+		}
+
+		if (!FindNextFile(hFind, &fd))
+			break;
+	}
+
+	if (hFind != INVALID_HANDLE_VALUE)
+		FindClose(hFind);
+
+	return false;
+
+}
+
+bool Directory::writeTimeHasChanged() const
+{
+	FILETIME newFiletime;
+
+	// if the new filetime can not be determined, return true for sure. The exact differences will be tested later.
+	if (!readLastWriteTime(newFiletime))
+		return true;
+
+	// we could successfully get the filetime, but before this it was 0 - definitely a change
+	if (!(_lastWriteTime.dwHighDateTime || _lastWriteTime.dwLowDateTime))
+		return true;
+
+	// compare file times
+	return _lastWriteTime.dwLowDateTime != newFiletime.dwLowDateTime
+		|| _lastWriteTime.dwHighDateTime != newFiletime.dwHighDateTime;
+
+}
+
+void Directory::setFilters(const std::vector<generic_string>& filters, bool autoread)
+{
+	if (_filters != filters)
+	{
+		_filters = filters;
+		if (autoread)
+			read();
+	}
+}
+
+void Directory::synchronizeTo(const Directory& other)
+{
+	onBeginSynchronize(other);
+	for (auto it = _dirs.begin(); it != _dirs.end(); ++it)
+		if (other._dirs.find(*it) == other._dirs.end())
+			onDirRemoved(*it);
+	for (auto it = other._dirs.begin(); it != other._dirs.end(); ++it)
+		if (_dirs.find(*it) == _dirs.end())
+			onDirAdded(*it);
+	for (auto it = _files.begin(); it != _files.end(); ++it)
+		if (other._files.find(*it) == other._files.end())
+			onFileRemoved(*it);
+	for (auto it = other._files.begin(); it != other._files.end(); ++it)
+		if (_files.find(*it) == _files.end())
+			onFileAdded(*it);
+	onEndSynchronize(other);
+
+	*this = other;
+
+}
+
+void Directory::setHideEmptyDirs(bool hideEmptyDirs, bool autoread)
+{
+	if (_hideEmptyDirs != hideEmptyDirs)
+	{
+		_hideEmptyDirs = hideEmptyDirs;
+		if (autoread)
+			read();
+	}
+}
+
+// read last write time reads the last write time into "filetime".
+// Under some circumstances, it might happen that the file time can not be determined, in this case it returns false.
+bool Directory::readLastWriteTime(FILETIME& filetime) const
```

Comment:
Is `filetime` intended to be an `_Out_` parameter? (yes, I think it is?) Mark it `_Out_`.

It's probably also a good idea (when marking it `_Out_`), to mark `Directory::readLastWriteTime` as such:

```
_Success_( return == true )
bool Directory::readLastWriteTime(_Out_ FILETIME& filetime) const
```
...so that static analysis can pick up misuse of the result (on a failed function call).